### ISSUE TITLE
Create a protobuf and/or Avro version of the ninjs schema

### ISSUE THREAD

**Issue by bquinn** [iptc/newsinjson#56]

**Comment by bquinn** [comment]
Tried using an npm script in today's call. Didn't work for arrays of objects - looks to be a bug in the npm module.

**Comment by bquinn** [comment]
Actually I found a patched version of the node.js module that works: https://github.com/Army-U/jsonschema-protobuf/

Here is the protobuf schema that it generates:

```
syntax = "proto3";

message Ninjs {
  message Person {
    optional string name = 1;
    optional string rel = 2;
    optional string scheme = 3;
    optional string code = 4;
  }
  
  message Organisation {
    message Symbols {
      optional string ticker = 1;
      optional string exchange = 2;
    }
    
    optional string name = 1;
    optional string rel = 2;
    optional string scheme = 3;
    optional string code = 4;
    repeated Symbols symbols = 5;
  }
  
  message Place {
    optional string name = 1;
    optional string rel = 2;
    optional string scheme = 3;
    optional string code = 4;
  }
  
  message Subject {
    optional string name = 1;
    optional string rel = 2;
    optional string scheme = 3;
    optional string code = 4;
  }
  
  message Event {
    optional string name = 1;
    optional string rel = 2;
    optional string scheme = 3;
    optional string code = 4;
  }
  
  message Object {
    optional string name = 1;
    optional string rel = 2;
    optional string scheme = 3;
    optional string code = 4;
  }
  
  message Infosource {
    optional string name = 1;
    optional string rel = 2;
    optional string scheme = 3;
    optional string code = 4;
  }
  
  message Renditions {
  }
  
  message Associations {
  }
  
  message Altids {
  }
  
  message Trustindicator {
    optional string scheme = 1;
    optional string code = 2;
    optional string title = 3;
    optional string href = 4;
  }
  
  message $standard {
    optional string name = 1;
    optional string version = 2;
    optional string schema = 3;
  }
  
  message Genre {
    optional string name = 1;
    optional string scheme = 2;
    optional string code = 3;
  }
  
  required string uri = 1;
  optional string type = 2;
  optional string mimetype = 3;
  optional string representationtype = 4;
  optional string profile = 5;
  optional string version = 6;
  optional string firstcreated = 7;
  optional string versioncreated = 8;
  optional string embargoed = 9;
  optional string pubstatus = 10;
  optional int32 urgency = 11;
  optional string copyrightholder = 12;
  optional string copyrightnotice = 13;
  optional string usageterms = 14;
  optional string ednote = 15;
  optional string language = 16;
  repeated Person person = 17;
  repeated Organisation organisation = 18;
  repeated Place place = 19;
  repeated Subject subject = 20;
  repeated Event event = 21;
  repeated Object object = 22;
  repeated Infosource infosource = 23;
  optional string title = 24;
  optional string byline = 25;
  optional string headline = 26;
  optional string slugline = 27;
  optional string located = 28;
  optional int32 charcount = 29;
  optional int32 wordcount = 30;
  optional Renditions renditions = 31;
  optional Associations associations = 32;
  optional Altids altids = 33;
  repeated Trustindicator trustindicator = 34;
  optional $standard $standard = 35;
  repeated Genre genre = 36;
}
```

So it seems that associations, renditions and altids will need some more work. As these are quite flexible properties and protobuf (as I understand it) is fairly locked down (that's why it's so simple and fast), we might not be able to support the same constructs in protobufs.

cc @viccro

**Comment by viccro** [comment]
This is great! I do recommend that we use [proto3](https://developers.google.com/protocol-buffers/docs/proto3), rather than [proto2](https://developers.google.com/protocol-buffers/docs/proto). I believe it's more widely used, and there were some nice upgrades to the language with the version change. 

I can take a look at patching the patched version to see if we can get that upgrade easily. This is a great start!

**Comment by viccro** [comment]
I'm admittedly not familiar with avro syntax, so I'd have to double check that this conversion works, but I used [this library](https://github.com/katsanva/jsonschema-avro) to create this first attempt at a schema:
```{ namespace: 'org.iptc.www.std.ninjs.ninjs.schema.1.3.json',
  name: 'http://www.iptc.org/std/ninjs/ninjs-schema_1.3.json#',
  type: 'record',
  doc:
   'A news item as JSON object -- copyright 2020 IPTC - International Press Telecommunications Council - www.iptc.org - This document is published under the Creative Commons Attribution 4.0 license, see  http://creativecommons.org/licenses/by/4.0/',
  fields:
   [ { name: 'uri',
       doc: 'The identifier for this news object. nar:newsItem@guid',
       type: 'string' },
     { name: 'type',
       doc:
        'The generic news type of this news object. (Value \'component\' added in version 1.2 as issue #21.). nar:itemClass',
       type: [Array],
       default: null },
     { name: 'mimetype',
       doc:
        'A MIME type which applies to this news object. nar:contentType',
       type: [Array],
       default: null },
     { name: 'representationtype',
       doc:
        'Indicates how complete this representation of a news item is. No mapping to nar. Specific for ninjs.',
       type: [Array],
       default: null },
     { name: 'profile',
       doc:
        'An identifier for the structure of the news object. This can be any string but we suggest something identifying the structure of the content such as \'text-only\' or \'text-photo\'. Profiles are typically provider-specific. nar:profile',
       type: [Array],
       default: null },
     { name: 'version',
       doc:
        'The version of the news object which is identified by the uri property. nar:newsItem@version',
       type: [Array],
       default: null },
     { name: 'firstcreated',
       doc:
        'Indicates when the first version of the item was created. (Added in version 1.2 from issue #5). nar:firstCreated',
       type: [Array],
       default: null },
     { name: 'versioncreated',
       doc:
        'The date and time when this version of the news object was created. nar:versionCreated',
       type: [Array],
       default: null },
     { name: 'embargoed',
       doc:
        'The date and time before which all versions of the news object are embargoed. If absent, this object is not embargoed. nar:embargoed',
       type: [Array],
       default: null },
     { name: 'pubstatus',
       doc:
        'The publishing status of the news object, its value is *usable* by default. nar:pubStatus',
       type: [Array],
       default: null },
     { name: 'urgency',
       doc:
        'The editorial urgency of the content from 1 to 9. 1 represents the highest urgency, 9 the lowest. nar:urgency',
       type: [Array],
       default: null },
     { name: 'copyrightholder',
       doc:
        'The person or organisation claiming the intellectual property for the content. nar:copyrightHolder',
       type: [Array],
       default: null },
     { name: 'copyrightnotice',
       doc:
        'Any necessary copyright notice for claiming the intellectual property for the content. nar:copyrightNotice',
       type: [Array],
       default: null },
     { name: 'usageterms',
       doc:
        'A natural-language statement about the usage terms pertaining to the content. nar:usageTerms',
       type: [Array],
       default: null },
     { name: 'ednote',
       doc:
        'A note that is intended to be read by internal staff at the receiving organisation, but not published to the end-user. (Added in version 1.2 from issue #6.) . ednote: nar:edNote',
       type: [Array],
       default: null },
     { name: 'language',
       doc:
        'The human language used by the content. The value should follow IETF BCP47. nar:language',
       type: [Array],
       default: null },
     { name: 'person',
       doc: 'An individual human being. nar:subject',
       type: [Array],
       default: null },
     { name: 'organisation',
       doc:
        'An administrative and functional structure which may act as as a business, as a political party or not-for-profit party. nar:subject',
       type: [Array],
       default: null },
     { name: 'place',
       doc: 'A named location. nar:subject',
       type: [Array],
       default: null },
     { name: 'subject',
       doc: 'A concept with a relationship to the content. nar:subject',
       type: [Array],
       default: null },
     { name: 'event',
       doc:
        'Something which happens in a planned or unplanned manner. nar:?',
       type: [Array],
       default: null },
     { name: 'object',
       doc: 'Something material, excluding persons. nar:subject',
       type: [Array],
       default: null },
     { name: 'infosource',
       doc:
        'A party (person or organisation) which originated, modified, enhanced, distributed, aggregated or supplied the content or provided some information used to create or enhance the content. (Added in version 1.2 according to issue #15.) .    infosource:  nar:infoSource',
       type: [Array],
       default: null },
     { name: 'title',
       doc:
        'A short natural-language name for the item. (Added in version 1.2 according to issue #9). nar:itemMeta/title',
       type: [Array],
       default: null },
     { name: 'byline',
       doc: 'The name(s) of the creator(s) of the content. nar:by',
       type: [Array],
       default: null },
     { name: 'headline',
       doc:
        'A brief and snappy introduction to the content, designed to catch the reader\'s attention. nar:headline',
       type: [Array],
       default: null },
     { name: 'slugline',
       doc:
        'A human-readable identifier for the item. (Added in version 1.2 from issue #4.). nar:slugline',
       type: [Array],
       default: null },
     { name: 'located',
       doc:
        'The name of the location from which the content originates. nar:located',
       type: [Array],
       default: null },
     { name: 'charcount',
       doc:
        'The total character count in the article excluding figure captions. (Added in version 1.2 according to issue #27.). nar:charcount',
       type: [Array],
       default: null },
     { name: 'wordcount',
       doc:
        'The total number of words in the article excluding figure captions. (Added in version 1.2 according to issue #27.). nar:wordcount',
       type: [Array],
       default: null },
     { name: 'renditions',
       doc:
        'Wrapper for different renditions of the news object. nar:remoteContent',
       type: [Array],
       default: null },
     { name: 'associations',
       doc:
        'Content of news objects which are associated with this news object. nar:link',
       type: [Array],
       default: null },
     { name: 'altids',
       doc:
        'Alternative identifiers of the item. It is up to the individual provider to name and set type on the alternative identifiers they like to use. nar:altId issue #3. (Added in version 1.3)',
       type: [Array],
       default: null },
     { name: 'trustindicator',
       doc:
        'An array of objects to allow links to documents about trust indicators. (nar:link) issue #44. (Added in version 1.3)',
       type: [Array],
       default: null },
     { name: '$standard',
       doc:
        'An object with information about standard, version and schema this instance is valid against. nar:standard, nar:standardversion and xml:schema issue #43. (Added in version 1.3)',
       type: [Array],
       default: null },
     { name: 'genre',
       doc:
        'A nature, intellectual or journalistic form of the content. nar:genre. (Added in version 1.3)',
       type: [Array],
       default: null } ] }
```
I only had to change "id" to "$id" to make the conversion work.

**Comment by bquinn** [comment]
Discussion on today's call:

* the way we use regexes doesn't work with avro and protobufs because everything must have an explicit name. So we could find a way of letting users extend the base schema with their own names that validate against our regex but are explicit for that user..?

* IDs of fields are based on the alphabetical order of property names when using the protobuf generator. So we could use the generator for the first version but then maintain it manually so we can guarantee backwards compatibility of the field identifiers in future versions.

* To enable multiple renditions, it would be best to include "rendition" as an object type (message) in protobufs, which means we would have to either change the structure of the JSON schema or to manually edit the generated protobuf version each time. (This could be another argument for going to 2.0 as the next version - it would let us make this change without worrying about it being a breaking change)

**Comment by bquinn** [comment]
I have updated the protobufs code now and we have a PR waiting to be merged into the 2.0 branch: #109 

@viccro are you back? Would you be able to take a look and let us know if you have any suggestions?
